src/peergos/server/corenode/IpfsCoreNode.java

```diff
@@ -0,0 +1,228 @@
+package peergos.server.corenode;
+
+import peergos.server.mutable.*;
+import peergos.shared.*;
+import peergos.shared.cbor.*;
+import peergos.shared.corenode.*;
+import peergos.shared.crypto.*;
+import peergos.shared.crypto.asymmetric.*;
+import peergos.shared.crypto.hash.*;
+import peergos.shared.hamt.*;
+import peergos.shared.io.ipfs.multihash.*;
+import peergos.shared.merklebtree.*;
+import peergos.shared.mutable.*;
+import peergos.shared.storage.*;
+import peergos.shared.user.*;
+import peergos.shared.util.*;
+
+import java.io.*;
+import java.net.*;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.*;
+import java.util.stream.*;
+
+public class IpfsCoreNode implements CoreNode {
+
+    public static final PublicSigningKey PEERGOS_IDENTITY_KEY = PublicSigningKey.fromString("ggFYIE7uD1ViM9KfiA1w69n774/jk6hERINN3xACPyabWiBp");
+    public static final PublicKeyHash PEERGOS_IDENTITY_KEY_HASH = PublicKeyHash.fromString("zdpuAvZynWLuyvovJwa34bj24M7cspt5M8seFfrFLrPWDGFDW");
+
+    private final ContentAddressedStorage ipfs;
+    private final MutablePointers mutable;
+    private final SigningPrivateKeyAndPublicHash signer;
+
+    private final Map<String, List<UserPublicKeyLink>> chains = new ConcurrentHashMap<>();
+    private final Map<PublicKeyHash, String> reverseLookup = new ConcurrentHashMap<>();
+    private final List<String> usernames = new ArrayList<>();
+
+    private MaybeMultihash currentRoot;
+
+    public IpfsCoreNode(SigningPrivateKeyAndPublicHash signer,
+                        MaybeMultihash currentRoot,
+                        ContentAddressedStorage ipfs,
+                        MutablePointers mutable) {
+        this.currentRoot = MaybeMultihash.empty();
+        this.ipfs = ipfs;
+        this.signer = signer;
+        this.mutable = mutable;
+        this.update(currentRoot);
+    }
+
+    private synchronized void update(MaybeMultihash newRoot) {
+        Consumer<Triple<ByteArrayWrapper, MaybeMultihash, MaybeMultihash>> consumer =
+                triple -> {
```

Comment:
Can you add a comment explaining what this is supposed to do (or extract into a method with a docstring)?